Migrates nll_loss_forward from TH to Aten (CUDA)

thomasjpfan opened the pull request:
Fixes https://github.com/pytorch/pytorch/issues/24610
Aten Umbrella issue https://github.com/pytorch/pytorch/issues/24507
Related to https://github.com/pytorch/pytorch/issues/59765

The performance does not change between this PR and master with the following benchmark script:

<details>
 <summary>Benchmark script</summary>

```python
import torch
import torch.nn as nn
import time

torch.manual_seed(0)


def _time():
    torch.cuda.synchronize()
    MS_PER_SECOND = 1000
    return time.perf_counter() * MS_PER_SECOND


device = "cuda"
C = 30
softmax = nn.LogSoftmax(dim=1)
n_runs = 250

for reduction in ["none", "mean", "sum"]:
    for N in [100_000, 500_000, 1_000_000]:
        fwd_t = 0
        bwd_t = 0
        data = torch.randn(N, C, device=device)
        target = torch.empty(N, dtype=torch.long, device=device).random_(0, C)
        loss = nn.NLLLoss(reduction=reduction)
        input = softmax(data)

        for i in range(n_runs):
            t1 = _time()
            result = loss(input, target)
            t2 = _time()
            fwd_t = fwd_t + (t2 - t1)
        fwd_avg = fwd_t / n_runs
        print(
            f"input size({N}, {C}), reduction: {reduction} "
            f"forward time is {fwd_avg:.2f} (ms)"
        )
    print()
```

</details>

## master

```
input size(100000, 30), reduction: none forward time is 0.02 (ms)
input size(500000, 30), reduction: none forward time is 0.08 (ms)
input size(1000000, 30), reduction: none forward time is 0.15 (ms)

input size(100000, 30), reduction: mean forward time is 1.81 (ms)
input size(500000, 30), reduction: mean forward time is 8.24 (ms)
input size(1000000, 30), reduction: mean forward time is 16.46 (ms)

input size(100000, 30), reduction: sum forward time is 1.66 (ms)
input size(500000, 30), reduction: sum forward time is 8.24 (ms)
input size(1000000, 30), reduction: sum forward time is 16.46 (ms)
```

## this PR

```
input size(100000, 30), reduction: none forward time is 0.02 (ms)
input size(500000, 30), reduction: none forward time is 0.08 (ms)
input size(1000000, 30), reduction: none forward time is 0.15 (ms)

input size(100000, 30), reduction: mean forward time is 1.80 (ms)
input size(500000, 30), reduction: mean forward time is 8.24 (ms)
input size(1000000, 30), reduction: mean forward time is 16.46 (ms)

input size(100000, 30), reduction: sum forward time is 1.66 (ms)
input size(500000, 30), reduction: sum forward time is 8.24 (ms)
input size(1000000, 30), reduction: sum forward time is 16.46 (ms)
```

thomasjpfan committed:
ENH Implements a10 version of nll_loss

thomasjpfan committed:
MAINT Migrates nllloss_forward to aten

thomasjpfan committed:
REV Reduces diff

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into nllloss_migration

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into nllloss_migration

thomasjpfan committed:
ENH Uses regular pointers

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into nllloss_migration

ezyang commented on aten/src/ATen/native/cuda/Loss.cu:
```diff
@@ -153,4 +154,290 @@ Tensor& binary_cross_entropy_backward_out_cuda(const Tensor& grad, const Tensor&
+  const int64_t n_dims = input.dim();
+  TORCH_CHECK(n_dims > 0 && n_dims <= 2, "input tensor should be 1D or 2D");
+  TORCH_CHECK(
+      target.dim() == 1,
```
What if target.dim() is 0? Previous code errored on this too

thomasjpfan replied:
I think the original implementation did not raise this specific error when `target.dim() == 0`:

https://github.com/pytorch/pytorch/blob/d5988c5eca0221e9ef58918e4f0b504940cb926a/aten/src/THCUNN/generic/ClassNLLCriterion.cu#L14-L16

I updated the check to `TORCH_CHECK(target.dim() <= 1)` to not change the original behavior.

ngimel replied:
Can you please beef up testing for supported/unsupported dims?

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into nllloss_migration

thomasjpfan committed:
STY Use consistent naming convention

thomasjpfan committed:
CLN Better match the order of the original implementation

thomasjpfan committed:
REV Straight port without change in behavior

thomasjpfan committed:
STY Better naming

thomasjpfan committed:
STY Better naming for threads

thomasjpfan committed:
STY Use full name for same gpu check

thomasjpfan committed:
ENH Closely match the original TH implementation

ngimel commented on aten/src/ATen/native/cuda/Loss.cu:
```diff
@@ -153,4 +154,290 @@ Tensor& binary_cross_entropy_backward_out_cuda(const Tensor& grad, const Tensor&
+
+  if (weight.defined()) {
+    TensorArg weight_arg{weight, "weight", 5};
+    checkAllSameGPU(
```
the device placement of all the args is automatically checked in codegen, you don't have to do it here (that also means you don't have to create TensorArgs

ngimel commented on aten/src/ATen/native/cuda/Loss.cu:
```diff
@@ -153,4 +154,290 @@ Tensor& binary_cross_entropy_backward_out_cuda(const Tensor& grad, const Tensor&
+      ")")
+
+  TORCH_CHECK(
+      !weight.defined() || weight.numel() == n_classes,
```
should you check weight.dim() here (and also add a test case for incorrectly sized/dimensioned weights?)

thomasjpfan replied:
I added more tests for input validation and also made a slight adjustment to the CPU version to also check for `weight.dim`.

ngimel commented on aten/src/ATen/native/cuda/Loss.cu:
```diff
@@ -153,4 +154,290 @@ Tensor& binary_cross_entropy_backward_out_cuda(const Tensor& grad, const Tensor&
+        [&] {
+          scalar_t* weight_data = nullptr;
+          if (weight.defined()) {
+            auto weight_ = weight.contiguous();
```
`contiguous` should be called outside of dispatch guard, and here you can just set `weight_data` with a ternary operator

ngimel commented on aten/src/ATen/native/cuda/Loss.cu:
```diff
@@ -153,4 +154,290 @@ Tensor& binary_cross_entropy_backward_out_cuda(const Tensor& grad, const Tensor&
+        [&] {
+          scalar_t* weight_data = nullptr;
+          if (weight.defined()) {
+            auto weight_ = weight.contiguous();
```
same here, move `weight.contiguous()` to outside of dispatch macro

ngimel reviewed (commented):
This looks good, I left a few comments

thomasjpfan committed:
ENH Address comments

thomasjpfan committed:
ENH Adds check for weights

thomasjpfan committed:
ENH Improve weights

thomasjpfan committed:
CLN Switch around logic

thomasjpfan committed:
REV Revert change to kernel

thomasjpfan committed:
CLN Use else

thomasjpfan commented on aten/src/ATen/native/cuda/Loss.cu:
```diff
@@ -153,4 +154,290 @@ Tensor& binary_cross_entropy_backward_out_cuda(const Tensor& grad, const Tensor&
+      output_acc += sh_inputs[i];
+      total_weight_acc += acc_weight[i];
+    }
+    *total_weight = static_cast<scalar_t>(total_weight_acc);
```
In this case, I think `*total_weight` still needs to be set here because it is one of the outputs of the kernel.

ngimel replied:
Ah, sorry, didn't notice that it's the output, you are right.

ngimel commented on aten/src/ATen/native/cuda/Loss.cu:
```diff
@@ -153,4 +154,260 @@ Tensor& binary_cross_entropy_backward_out_cuda(const Tensor& grad, const Tensor&
+      }
+      if (total_weight_acc != 0) {
+        *output = static_cast<scalar_t>(output_acc / total_weight_acc);
+      }
```
you need `else` here, otherwise `output` won't be set for total_weight_acc = 0, and please add tests for all these cases.

thomasjpfan replied:
The original implementation had `*output` initialized to 0, which happened to be correct when `total_weight_acc==0`.

I redid the implementation a little to only set `*output` once and added test for the `nan` case and the `size_average && total_weight_acc == 0` case.

thomasjpfan committed:
ENH Adds test for zero weights and improves flow

thomasjpfan committed:
TST Adds test for nan

ngimel commented:
Test errors are real

thomasjpfan committed:
CLN Typo

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into nllloss_migration

thomasjpfan committed:
REV Remove diff

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into nllloss_migration

thomasjpfan committed:
BLD Completely remove nll_loss

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into nllloss_migration_rb

thomasjpfan committed:
REV Less diffs

thomasjpfan committed:
BLD Fix merge

thomasjpfan committed:
REV Less diffs

thomasjpfan committed:
BLD Less diffs

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into nllloss_migration

thomasjpfan committed:
TST Removes unneeded tests

thomasjpfan commented:
While looking at a [test error](https://app.circleci.com/pipelines/github/pytorch/pytorch/340276/workflows/8b476edc-b515-44b2-8420-2e4b8171220a/jobs/14312737) I saw that we already have test for zero weights and empty tensors: `test_nll_loss_empty_tensor_reduction_*` and `test_nll_loss_total_weight_is_zero`, so I removed the tests I added in this PR that tested the same thing.

The rest of the public tests seems unrelated to this PR.
